clients/csharp-dotnet-core/voice-assistant-test/tool/BotConnector.cs

```diff
@@ -600,11 +611,13 @@ private void SpeechBotConnector_Canceled(object sender, SpeechRecognitionCancele
 
         private void SpeechBotConnector_SessionStarted(object sender, SessionEventArgs e)
         {
+            this.stopWatch.Start();
             Trace.TraceInformation($"[{DateTime.Now.ToString("h:mm:ss tt", CultureInfo.CurrentCulture)}] Session Started event received. SessionId = {e.SessionId}");
         }
 
         private void SpeechBotConnector_SessionStopped(object sender, SessionEventArgs e)
         {
+            this.stopWatch.Stop();

```

Comment:
The stopwatch should start when Speech SDK starts consuming audio from the input stream. We use the session started event as an approximation of that. The stopwatch should stop when the first TTS audio buffers is received (if there is TTS audio), or when the bot-response activity is received (if it does not have TTS audio attached).
The UPL value is then:
StopTime - StartTime + LenghtOfSpeechInWavFiles.
where LenghtOfSpeechInWavFiles is the new field you just addded and called WavFileDuration (we need to agree on a name).
If the field LenghtOfSpeechInWavFiles  (aka WavFileDuration) was not specified, than we take the duration of the whole input WAV file. Which is equivalent to saying if you did  not tell us when speech ended in the WAV file, we assume it right at the end of the WAV file.